### DISCUSSION TITLE
Automatic Fallback

### DISCUSSION THREAD

**Discussion by narendasan** [pytorch/TensorRT#360] [RFCs]
# Automatic Fallback

**Goal**: Allow the compiler to identify subgraphs that can be supported by TRTorch and correctly segment out these graphs, compile each engine and then link together TorchScript and TRTorch engines. 

Currently as when users hit an unsupported op, compilation is aborted. Users have requested that in the event of an unsupported op that it would be run in PyTorch. This wildly increases the complexity of the compiler. PyTorch also operates on a JIT basis, not requiring ahead of time understanding of metadata such as shape or dependencies. Since it executes in topological order while traversing the graph, correct execution is guaranteed. We therefore in order to implement fallback will need to determine which subgraphs we can handle without interfering with the execution order and also ensuring all ahead of time information is available. The goal is to be able to return to users a new graph which has both PyTorch ops and TensorRT ops and still ensure correct execution. 

For example, the following forward function could be partitioned like so: 

```python
def forward(dict: Dict[str, Any]):
   z = F.max_pool(self.conv(dict["in"])) # Tensor Computation [TRT] 
   z1 = dict["scale_1"] * dict["scale_2"] # Scalar [@ RT]
   y = z * z1 # Scalar * Tensor [@ RT] 
   z = Backbone(y) # [TRT]  
   return z
```

The scalar operations and dictionary accesses are handled by Torch but the Tensor computation is done by TensorRT. 

## Usecase for Fallback

Fallback is useful if the following conditions cannot be met:

1. Unsupported ops are so common that segmenting models based on support for TRT explodes the codebase
2. Writing converters for TRTorch is not possible or there is too many to write

## Proposed User-Level API 

There are a few settings that we would want to support

```python
import torch
import trtorch

...
model = MyModel()
ts_model = torch.jit.script(model)
trt_model = trtorch.compile(model, {
  ...
  "torch_fallback" : {
    "enabled" : True,
    "min_block_size" : 3,
    "forced_fallback_operators": ["aten::foo"],
  } 
})
```

By default fallback will be off, it is enabled by setting `enabled` as `True`. Next there will be specification for a minimum block size. What this means is that if there is a series of in this example at least 3 consecutive operations that can be converted to TensorRT then an engine will be created to encapsulate those operations. Finally there is a list of strings that will be names of operations that the user explicitly wants to run in PyTorch.

### Lowering

Current lowering used today will run prior to anything related to fallback. This will make sure we are working with the most TensorRT compatible graph we can be.

### Partitioning 

The first stage of fallback will be the partitioning stage. Here using the settings provided by the user, the graph will be segmented into parts that are going to be in TensorRT and parts that will remain in PyTorch. 

#### Core Data Structures

We need to maintain information about the structure of the graph after partitioning that will allow us to reconstruct the graph once conversion of all TensorRT subgraphs is done. My suggestion is a structure like the following: 

```c++
struct SegmentedBlock {
  enum SegmentedBlockTarget {
  	kTorch,
  	kTensorRT,
	};
  
  SegmentedBlockTarget target; //Whether this section of the graph is going to be in TensorRT or LibTorch
  nvinfer1::Dims or c10::List in_shape; // Needs to support dynamic shape 
  nvinfer1::Dims or c10::List out_shape;  // Needs to support dynamic shape
  std::vector<torch::jit::Value*> inputs;
  std::vector<torch::jit::Value*> outputs;
  std::vector<torch::jit::Node*> nodes; //This could maybe also be torch::jit::Block, not sure if there advantage here for doing that 
  std::string trt_engine;
};
```

The target defines if the section of the graph will remain in Torch or TensorRT. Input and output shape are either provided or calculated through the course of partitioning. Inputs are a list of ` Value` pointers which define the inputs to this section of the graph, Similarly with the outputs. As we traverse the list of segmented blocks we should be able to now reconstruct the connections between blocks. Nodes are the list of nodes in the block, stored topologically. This in the case the blocks targets TensorRT will be fed directly into the conversion system otherwise this will be used to reconstruct the subgraph in the final synthesized program. Finally there is a field for the serialized TensorRT engine if applicable. 

 The partitioning system will go through the graph building these blocks and storing them sorted topologically in an ordered list.

#### Tensor Shape Calculation

As TRTorch operates now, we are given the input shape for the model. As we traverse the graph we can maintain a running calculation of the current tensor shape. Now that we are not necessarily guaranteed to have input shapes for each subgraph requested to be compiled, we need to explicitly calculate tensor shapes as we traverse the graph so that when we go to compile a subgraph we have the input shape for the engine. This is going to be one of the hardest parts of this work since we do not have a fully exhaustive library to calculate Tensor shape. We can attempt to run the graph provided or the subgraphs themselves through JIT with dummy data during compilation to calculate these sizes as well. In the case of dynamic shape we would need to do this through the input range provided perhaps. 

##### Fundamental Assumptions 

1. Inputs to a TRT block will have a standard interface which is it consumes a `Tensor` or List of `Tensor`s and produces only a `Tensor` or List of `Tensor`s
2. TRT blocks only encapsulate operations that can be formalized as a pure function: `Tensor or [Tensor] -> Tensor or [Tensor]`. TensorRT cannot handle side effects. 

##### Steps

1. Verifying Op Support 

   - Go through the set of operators in order and mark those which do and do not have converters. We have `torch::core::conversion::OpSupported` to help here and we also must respect the `forced_fallback_operators` from the user. 

2. Preliminary Partitioning 

   - At this point we can roughly separate the graph into parts that we can support and parts we cannot. At this stage we can start to create the first blocks that will target libtorch (more will be created later as we trim things off the compilation set but it is not important that we minimize the number of blocks).
   - Interfaces (inputs and outputs) for torch blocks need to be defined so that when we go to look at TensorRT blocks we can know that these block outputs can be treated as input ` ITensors` from the TensorRT perspective later.

3. Calculate Dependencies 

   - For every operator that we could potentially support we need to verify that all input data to the operator can be resolved at compile time given that some data may be run through libtorch hence only will be available at runtime.

     - For every to be compiled operator there must be a "complete dependency graph". This means that every input can be traced back to an input `ITensor` (i.e. is either the input to the graph or is an output of a previously identified torch block) or is or a graph constant. Otherwise this op cannot be supported. There is no case where if a op takes an input that is anything but a Tensor and it is not a constant where we can support it as of now due to the fundamental assumptions 
     - Those ops that have incomplete dependency graphs will be removed from the compilation set and a new torch block should be created for that op

   - Edge cases (Non exhaustive):

     - There is a supported operator that has two consumers of its output, one block is supported and the other is not.

       ![img](https://docs.google.com/drawings/u/0/d/skhZcuolYV29w_dLQD5tVZQ/image?w=341&h=290&rev=60&ac=1&parent=1wp6q3fXu-YlgMHWBWQHotNtTtuLTDCK4dw09FhgVqXg)

       - There are a couple options:
         - Don’t include CONV in TRT engine Backbone1 and run it in PyTorch
         - Include CONV in Backbone1 engine but have an additional output for CONV1 to feed into Backbone2
         - Make CONV a separate block subject to `min_block size `

     - There is a supported operator that consumes an output from a supported block and an unsupported block ![img](https://docs.google.com/drawings/u/0/d/sdyjYfwVHirZ-kRGlllB3Ug/image?w=341&h=297&rev=12&ac=1&parent=1wp6q3fXu-YlgMHWBWQHotNtTtuLTDCK4dw09FhgVqXg)

       - [Conservative] Don't compile ADD and run it in PyTorch
       - Have two inputs to the graph for backbone 1 and ADD 
         - We then must ensure that the execution order forces Backbone 2 to run first at runtime, this may require a dummy input 
       - Make ADD a separate block subject to `min_block_size ` 

4. Prune TensorRT blocks according to number of nodes

   1. Once the compilation set has been created verifying all necessary data will be available for each TensorRT block, check the size of each block and those which are too small get converted to Torch Blocks

5. Tensor Shape Calculation

   1. For each block, figure out the input and output shapes starting from the provided input shape from the user. This can be done through running the graph with JIT perhaps. For dynamic shape multiple runs may be required   

### Conversion 

For conversion we iterate through the set of identified TensorRT blocks, for each of them running our current day conversion phase, but using the subgraph and input shape defined in the block struct producing a serialized TensorRT engine.

### Program Synthesis 

From this point we now have to reconstruct the program using the pieces from both LibTorch and TensorRT in a single new module. Using the list of blocks we can now reconstruct a copy of the program in a new module. The graph interface should be the same as the input graph. For each TensorRT Block, the subgraph will be the following series of operations:

```
  %1 : __torch__.torch.classes.tensorrt.Engine = prim::GetAttr[name="__torch___torchvision_models_alexnet____torch_mangle_54_AlexNet_trt_engine"](%self_1)
  %3 : Tensor[] = prim::ListConstruct(%input_0)
  %4 : Tensor[] = tensorrt::execute_engine(%3, %1)
  %5 : Tensor = prim::ListUnpack(%4)
```

Where `%1` is the engine itself, `%3` aggregates inputs to the TensorRT engine (this should be all input values to the block) and `%5` disaggregates the outputs (corresponding to the output values of the block). Once each block has been added, the module can be returned to the user.

**Comment by narendasan** [comment]
cc: @peri044 @bowang95

**Comment by narendasan** [comment]
I think actually we might be able to use the output shape calculation to get intermediate layer shapes similar to how we do now. We will still need to explicitly pull them out and store them and it's still a bit hard to reason about the dynamic shape case but I think it will be the same as the static just with a bunch of placeholders. The runtime will need to ensure that it fills in the tensor sizes for each of the intermediate inputs

**Comment by narendasan** [comment]
We should make sure that we store some information about what each engine does in the final graph so people dumping the graph can see what got compiled and what didn't.  https://github.com/NVIDIA/TRTorch/issues/47

**Comment by inocsin** [comment]
Hi Bo, I tested some models with fallback, most of them ended with this error

```
unknown file: Failure
C++ exception with description "isTensor() INTERNAL ASSERT FAILED at "bazel-out/k8-fastbuild/bin/external/libtorch/_virtual_includes/ATen/ATen/core/ivalue_inl.h":130, please report a bug to PyTorch. Expected Tensor but got Int
Exception raised from toTensor at bazel-out/k8-fastbuild/bin/external/libtorch/_virtual_includes/ATen/ATen/core/ivalue_inl.h:130 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x69 (0x7fc1663fcb89 in /home/vincentzh/Projects/TRTorch/bazel-TRTorch/external/libtorch/lib/libc10.so)
frame #1: c10::IValue::toTensor() const & + 0xef (0x55ee2bdb507b in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/test_fallback.runfiles/TRTorch/tests/core/conversion/converters/test_fallback)
frame #2: trtorch::core::partitioning::registerSegmentInOutShape(trtorch::core::partitioning::SegmentedBlock&, std::unordered_map<torch::jit::Value*, nvinfer1::Dims, std::hash<torch::jit::Value*>, std::equal_to<torch::jit::Value*>, std::allocator<std::pair<torch::jit::Value* const, nvinfer1::Dims> > >&) + 0x933 (0x7fc1ed12d5fd in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libcore_Spartitioning_Slibpartitioning.so)
frame #3: trtorch::core::partitioning::segment_graph(std::shared_ptr<torch::jit::Graph>, std::vector<trtorch::core::conversion::InputRange, std::allocator<trtorch::core::conversion::InputRange> >&, trtorch::core::conversion::TorchFallback const&) + 0x52b (0x7fc1ed12e65c in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libcore_Spartitioning_Slibpartitioning.so)
frame #4: trtorch::core::CompileGraphWithFallback(torch::jit::Module const&, trtorch::core::CompileSpec) + 0x3d2 (0x7fc1ef888dad in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libcore_Slibcore.so)
frame #5: trtorch::core::CompileGraph(torch::jit::Module const&, trtorch::core::CompileSpec) + 0x78 (0x7fc1ef889a2b in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libcore_Slibcore.so)
frame #6: <unknown function> + 0x34901 (0x55ee2bdc4901 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/test_fallback.runfiles/TRTorch/tests/core/conversion/converters/test_fallback)
frame #7: <unknown function> + 0x2250e (0x55ee2bdb250e in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/test_fallback.runfiles/TRTorch/tests/core/conversion/converters/test_fallback)
frame #8: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x65 (0x7fc1d093ba72 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #9: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x5a (0x7fc1d093725b in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #10: testing::Test::Run() + 0xee (0x7fc1d0921648 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #11: testing::TestInfo::Run() + 0x10f (0x7fc1d0921fa9 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #12: testing::TestSuite::Run() + 0x12c (0x7fc1d0922680 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #13: testing::internal::UnitTestImpl::RunAllTests() + 0x41a (0x7fc1d092df6a in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #14: bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x65 (0x7fc1d093caf5 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #15: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x5a (0x7fc1d09380ff in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #16: testing::UnitTest::Run() + 0xc9 (0x7fc1d092c8dd in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #17: RUN_ALL_TESTS() + 0x11 (0x7fc1d0982aa8 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest_Umain.so)
frame #18: main + 0x3d (0x7fc1d0982a37 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest_Umain.so)
frame #19: __libc_start_main + 0xe7 (0x7fc16544cb97 in /lib/x86_64-linux-gnu/libc.so.6)
frame #20: <unknown function> + 0x2180a (0x55ee2bdb180a in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/converters/test_fallback.runfiles/TRTorch/tests/core/conversion/converters/test_fallback)
" thrown in the test body.
```

**Reply by bowang007** [reply]
It seems that the code broke down in the shape inference process, one of the input that passed in for segmented graph is not Tensor I guess, could you please share with us the model? Thanks!

**Comment by peri044** [comment]
This is the test suite which can cover all the components of the automatic fallback feature.

# Test suite

Each testcase would take an simple input graph with atleast one unsupported node.

## 1) segmentation

fallback base model conv1-conv2 - log_sig -> conv3

a) assert the number of segments  to be correct 2 trt segments, 1 pyt segment  

b) Also check if the segments have the right nodes. For ex: 1st trt segment should have 2 conv (conv1 + conv2) nodes, pytorch segment should have log_sig node and 2nd trt segment should have conv3 layer 

c) min_seg_size : 3 , no: segments for trt : 0, for pyt : 4

d) forced_fallback operator test

### 2) Check the validity of mini-graph - check if the nodes in mini-graph are present in global lowered graph through an unordered map.

Iterate through mini-graphs and ensure all the nodes exist

### 3) shape inference / mini-graph execution (after checking each of them)

This step is to ensure all the shapes of a mini-graph are known. 
In your example, you should have two blocks and hardcoded way to check the shapes with the shape inference. 

### 4) TensorRT conversion of mini-graphs

conv-conv - log_sig -> conv with min_seg_size=2
Compare how many TRT engines in graph: in this example : 1 trt engine. 
3 modules - In one module, trt engine will be wrapped. Check this in the assert statement. 

### 5) stitching mini-graphs

For every node in the new global fallback graph, there would be a corresponding input in the original graph. 
2 block sample network and can we ensure block 1 outputs are block2's inputs ? Verify this. 

### 6) end-end run of hybrid graph vs pytorch graph run 

build a hybrid graph and run 
a) pytorch original graph

b) fallback optimized graph

Compare the outputs.

**Comment by narendasan** [comment]
There is a question of if users use truncation and automatic fallback together, if weights get truncated in a TensorRT block and a following PyTorch block expects a tensor that is Float64 or Long, what should be the correct semantics so that we dont get type errors? Does the TensorRT block need to track uses of truncated tensors so it can cast back at the end?

**Comment by bowang007** [comment]
currently more corner cases are covered:
1. When running PyTorch segments, all kinds of outputs' IValues will be recorded rather than only output tensor shapes. This is necessary because later segments may have inputs not just Tensors.
2.  For PyTorch segments, almost all kinds of inputs and outputs are supported rather than only Tensor inputs/outputs for TensorRT segments.
3. For TensorRT segments, there might be some cases that the input/output is not Tensor. The solution is either fallback the whole segment to PyTorch or support some other types' input/outputs for TensorRT. Currently, types such as Int/Bool are supported by inserting prim::constants into the segments to achieve the similar functionalities.

**Comment by peri044** [comment]
The feature should have the following principle post segmentation. 
### Dependency resolution in segmentation

We would need all TRT segments in the graph to be self-sustainable. This requires duplicating the dependency chain of the inputs in TRT segments.

Eg: Input graph

```
//  Assume this part is being used by following TRT and pytorch blocks
%1 : int = prim::Constant[value=1]()
%2 : int[] = prim::ListConstruct(%1, %1)

%3 : Tensor = aten::max_pool(%input_1, %2) // Should be converted to TRT segment
.
.
%50 : Tensor = aten::pytorch_op(%input_49, %2) // Fallback to pytorch
```

The output graph should look like 

```
# Leaving these nodes as is in the pytorch graph for pytorch layers to consume
%1 : int = prim::Constant[value=1]()
%2 : int[] = prim::ListConstruct(%1, %1)

// This is a self-sustained graph with all dependencies within the block
%1_copy_1: int = prim::Constant[value=1]()
%2_copy_1: int[] = prim::ListConstruct(%1_copy_1, %1_copy_1)
%3 : Tensor = aten::max_pool(%input_1, %2_copy_1)
.
.
// This is a pytorch segment which consumes the original %2
%50 : Tensor = aten::pytorch_op(%input_49, %2)
```



We should make sure the entire dependency chain is duplicated and not just inserting the value of %2 directly in the TRT segment inputs.
